Assets/MixedRealityToolkit/Definitions/Devices/MixedRealityInteractionMapping.cs

```diff
@@ -401,22 +401,9 @@ public Vector2 Vector2Data
                     Debug.LogError($"SetVector2Value is only valid for AxisType.DualAxis InteractionMappings\nPlease check the {inputType} mapping for the current controller");
                 }
 
-                if (invertXAxis || invertYAxis)
-                {
-                    float invertXAxisFactor = invertXAxis ? -1f : 1f;
-                    float invertYAxisFactor = invertYAxis ? -1f : 1f;
-
-                    Changed = !vector2Data.x.Equals(value.x * invertXAxisFactor) &&
-                              !vector2Data.y.Equals(value.y * invertYAxisFactor);
-
-                    vector2Data.x = value.x * invertXAxisFactor;
-                    vector2Data.y = value.y * invertYAxisFactor;
-                }
-                else
-                {
-                    Changed = vector2Data != value;
-                    vector2Data = value;
-                }
+                Vector2 newValue = value * new Vector2(invertXAxis ? -1f : 1f, invertYAxis ? -1f : 1f);
```

Comment:
creating a new `Vector2` here will impact performance.

Reply:
Yep - I believe this gets referenced in the update loop and will end up leading to small allocs (which then trigger GC down the line). There's a couple possibilities here:

1) Switch to ||.

Advantages being less churn, but also possibly less readable code (i.e. it's easy to look at vector2Data != newValue, still easy to read the above but a little less natural)

2) Have a class-member variable for temp space to avoid alloc.

Slightly readable but now we have a what like 16 byte increase in class memory usage.

I don't really have a preference.

---
In reply to: [279413081](https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/4116#discussion_r279413081) [](ancestors = 279413081)

Reply:
I'm quite ignorant about the life-cycle of temporary value-type objects in C#, but I would have expected the compiler being able to do something reasonable there. Can you elaborate a bit more on where that performance impact comes from?

Reply:
> 1. Switch to ||.

💯 that was also my original suggestion.

> I would have expected the compiler being able to do something reasonable there

That's partly why we're being pretty explicit about how we're handling these value types here.

> Can you elaborate a bit more on where that performance impact comes from

This entire class was written to be highly efficient because it's **_very_** hot code, and we were trying to make sure all the memory was packed in tight. You're creating a `new` reference type that will need to be collected later and throws off the tightly packed memory layout. Why waste the time writing to a new memory address, when the one you have works just fine?

Reply:
This makes me think that the Vector2 creation is not that expensive, but I'd be super happy if someone has some more info about what code ends up being generated.

> Both value-type objects such as structs and reference-type objects such as classes are destroyed automatically, but value-type objects are destroyed when their containing context is destroyed, whereas reference-type objects are destroyed by the garbage collector at an unspecified time after the last reference to them is removed

From https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/new-operator.

Reply:
> You're creating a new reference type

As far as I can tell Vector2 is a value type. Are you refering to a different type or am I missing something?

Roughly, how many times per frame do you expect this setter to be called? Do you have a use case in mind where this code path would be specially hot?

Reply:
Sorry, you're correct. value type. I updated my comment to reflect.

All of the input mapping code is hot. Input has and always will be a performance critical area.

Reply:
These perf concerns come up often enough that I think a burden-of-proof approach is warranted. Otherwise they just turn into a (usually unresolved) back-and-forth about hypotheticals.

If we think this approach will cause a measurable performance problem, great - measure it. Profile and/or stopwatch both approaches and post the results. If the difference is negligible, discussion over. If it isn't, discussion over.

Reply:
I mean @SimonDarksideJ went back and forth for weeks testing the perf, and I'm pretty confident there's results in the GitHub history somewhere in here.

Let's keep it the way it is and make minimal changes to fix the existing code because we know, without a shadow of a doubt it works and is performance tested.

IMHO The burden of proof falls on the one opening the PR making these kinds of sweeping changes.

But then again, it's not like you guys are keen on listening to us anyway, so at this point, I'm not even sure why I bother anymore. Your loss.

Reply:
Don't misunderstand me - I concede there could be an issue here. My suggestion is more about reducing the mental load of discussing it. Plus for hot code like this we should really be in the habit of profiling anyway.

Reply:
The conversation really went on here, but I wanted to mention:

1) Unity's own guidelines recommend against allocations in Update/LateUpdate (i.e. inner loop stuff)

https://unity3d.com/learn/tutorials/topics/performance-optimization/optimizing-garbage-collection-unity-games?playlist=44069

See "Don’t allocate in functions that are called frequently"

2) Once things go down to IL2CPP, I have no idea if any of the .NET rules for GC still apply given that it's some unity IL2CPP stuff that's going on. Maybe it's smart enough to actually cache that new value once (and thus it will keep using the same thing on the heap with no issues). The problem is that we don't know (or if someone does know and can point to data, please do!)

3) Given Unity's own recommendation, if we trust that Unity's recommendation is reasonable (which I believe is the case) and we don't yet have the cycles to run our own profiling to double check it, it seems reasonable to err on the side of following Unity's performance guidelines in how we approach our problems.